waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+
+  # Since the EC multiplication writes the result to the input, we copy the input to the output variable 
+  output = public
+  # We execute the DH operation
+  EllipticCurve.mul(output, private)
+
+  return output
+
 
 #################################################################
 
+# Noise state machine primitives
+
+# Overview :
+# - Alice and Bob process (i.e. read and write, based on their role) each token appearing in a handshake pattern, consisting of pre-message and message patterns;
+# - Both users initialize and update according to processed tokens a Handshake State, a Symmetric State and a Cipher State;
+# - A preshared key psk is processed by calling MixKeyAndHash(psk);
+# - When an ephemeral public key e is read or written, the handshake hash value h is updated by calling mixHash(e); If the handshake expects a psk, MixKey(e) is further called
+# - When an encrypted static public key s or a payload message m is read, it is decrypted with decryptAndHash;
+# - When a static public key s or a payload message is writted, it is encrypted with encryptAndHash;
+# - When any Diffie-Hellman token ee, es, se, ss is read or written, the chaining key ck is updated by calling MixKey on the computed secret;
+# - If all tokens are processed, users compute two new Cipher States by calling Split;
+# - The two Cipher States obtained from Split are used to encrypt/decrypt outbound/inbound messages.
+
+#################################
+# Cipher State Primitives
+#################################
+
+# Checks if a Cipher State has an encryption key set
+proc hasKey(cs: CipherState): bool =
+  return (cs.k != EmptyKey)
+
+# Encrypts a plaintext using key material in a Noise Cipher State
+# The CipherState is updated increasing the nonce (used as a counter in Noise) by one
+proc encryptWithAd*(state: var CipherState, ad, plaintext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseNonceMaxError].} =
+  
+  # The output is the concatenation of the ciphertext and authorization tag
+  # We define its length accordingly
+  var ciphertext = newSeqOfCap[byte](plaintext.len + sizeof(ChaChaPolyTag))
+  
+  # Since ChaChaPoly encryption primitive overwrites the input with the output,
+  # we copy the plaintext in the output ciphertext variable and we pass it to encryption
+  ciphertext.add(plaintext)
+
+  # The nonce is read from the input CipherState
+  # By Noise specification the nonce is 4 bytes long out of the 12 bytes supported by ChaChaPoly
+  var nonce: ChaChaPolyNonce
+  nonce[4..<12] = toBytesLE(state.n)
+
+  # We perform encryption and we store the authorization tag
+  var authorizationTag: ChaChaPolyTag
+  ChaChaPoly.encrypt(state.k, nonce, authorizationTag, ciphertext, ad)
+
+  # We append the authorization tag to ciphertext
+  ciphertext.add(authorizationTag)
+
+  # We increase the Cipher state nonce
+  inc state.n
+  # If the nonce is greater than the maximum allowed nonce, we raise an exception
+  if state.n > NonceMax:
```

Comment:
Wondering why the `NonceMax+1` i.e., `uint64.high-1+1` value is an invalid nonce? it can still be used for the next message? I think the nonce needs to be check before increment i.e., to see if it is equal to `uint64.high`, and then avoid increment and throw an exception indicating the max value is hit. 
Ditto for the similar nonce max check in the `decryptWithAd` proc. https://github.com/status-im/nwaku/blob/795eefcd697cbae4e0039496c39544383682500a/waku/v2/protocol/waku_noise/noise.nim#L429

Reply:
I agree with your observation! The reason is that, by specification, 2^64-1 is a reserved nonce. I quote verbatim:

> A CipherState responds to the following functions. The ++ post-increment operator applied to n means "use the current n value, then increment it". The maximum n value (2^64-1) is reserved for other use. If incrementing n results in 2^64-1, then any further EncryptWithAd() or DecryptWithAd() calls will signal an error to the caller.

more on [The Cipher State Object](http://www.noiseprotocol.org/noise.html#the-cipherstate-object).

Note that `NonceMax = uint64.high - 1`, that is `NonceMax = 2^64-1-1`. Hence, if `n > NonceMax` after an increment, it means that `n = NonceMax+1 = 2^64-1`. In fact, `NonceMax` is the last "valid" nonce that can be used in encryption/decryption according to specification.

Reply:
thanks, clarified!